File: .scratch/memory-stage-4/148-owner-frozen-files/internal/memory/compiler_diagnostics.go
package memory

// Diagnostics contains operational metadata only. Neither an owner approval nor
// elapsed inbox age measures extraction accuracy or active human review time.
const CompilerDiagnosticsMaxPage = 32

type CompilerDiagnosticsQuery struct {
	SessionID    SessionID `json:"session_id"`
	View         string    `json:"view"`                    // jobs, candidates, activations, history, selections, live_roots, selection, foreground
	GenerationID string    `json:"generation_id,omitempty"` // required for selection
	Cursor       string    `json:"cursor,omitempty"`
	Limit        int       `json:"limit,omitempty"`
}

type CompilerDiagnosticSessionQuery struct {
	Cursor string `json:"cursor,omitempty"`
	Limit  int    `json:"limit,omitempty"`
}

type CompilerDiagnosticSessions struct {
	SessionIDs []SessionID `json:"session_ids"`
	NextCursor string      `json:"next_cursor,omitempty"`
}

type CompilerDiagnostics struct {
	Selections    []CompilerDiagnosticUnit        `json:"selections"`
	LiveRoots     []CompilerDiagnosticRoot        `json:"live_roots"`
	ScopeKey      string                          `json:"scope_key"`
	SessionID     SessionID                       `json:"session_id"`
	View          string                          `json:"view"`
	AsOfUnixMS    int64                           `json:"as_of_unix_ms"`
	Revision      int64                           `json:"revision"`
	Indexing      bool                            `json:"indexing"`
	Counts        map[string]int64                `json:"counts"`
	CapacityState string                          `json:"capacity_state"`
	Jobs          []CompilerDiagnosticJob         `json:"jobs"`
	Candidates    []CompilerDiagnosticCandidate   `json:"candidates"`
	Activations   []CompilerActivation            `json:"activations"`
	History       []CompilerDiagnosticHistory     `json:"history"`
	Selection     []CompilerDiagnosticSelection   `json:"selection"`
	Foreground    []CompilerForegroundMeasurement `json:"foreground"`
	NextCursor    string                          `json:"next_cursor,omitempty"`
}

type CompilerDiagnosticJob struct {
	PublicationNanos        *int64 `json:"publication_nanos"`
	CandidateFreshnessNanos *int64 `json:"candidate_freshness_nanos"`
	CompilerWorkStatus
	QueuedAtUnixMS     *int64                       `json:"queued_at_unix_ms"`
	PublishedAtUnixMS  *int64                       `json:"published_at_unix_ms"`
	SelectedNewEvents  int64                        `json:"selected_new_events"`
	CompletedNewEvents int64                        `json:"completed_new_events"`
	Measurements       []CompilerAttemptMeasurement `json:"measurements"`
}

type CompilerAttemptMeasurement struct {
	Attempt                 int    `json:"attempt"`
	Fence                   int64  `json:"fence"`
	ClaimedAtUnixMS         int64  `json:"claimed_at_unix_ms"`
	QueueWaitNanos          *int64 `json:"queue_wait_nanos"`
	InferenceNanos          *int64 `json:"inference_nanos"`
	ValidationNanos         *int64 `json:"validation_nanos"`
	DatabaseCompletionNanos *int64 `json:"database_completion_nanos"`
	ObservedOutcome         string `json:"observed_outcome"` // incomplete, completed, failed, cancelled, stale
}

type CompilerDiagnosticCandidate struct {
	Ref               CandidateRef `json:"ref"`
	JobID             string       `json:"job_id"`
	GenerationID      string       `json:"generation_id"`
	ReviewState       string       `json:"review_state"`
	EquivalentTo      string       `json:"equivalent_to,omitempty"`
	PublishedAtUnixMS *int64       `json:"published_at_unix_ms"`
	DecidedAtUnixMS   *int64       `json:"decided_at_unix_ms"`
	Edited            bool         `json:"edited"`
}

type CompilerDiagnosticHistory struct {
	RequestID       string `json:"request_id"`
	RangeIndex      int    `json:"range_index"`
	GenerationID    string `json:"generation_id"`
	FirstSequence   int64  `json:"first_sequence"`
	LastSequence    int64  `json:"last_sequence"`
	ScannedSequence int64  `json:"scanned_sequence"`
	Revision        int64  `json:"revision"`
	Cancelled       bool   `json:"cancelled"`
}

type CompilerDiagnosticSelection struct {
	EventID    EventID `json:"event_id"`
	Sequence   int64   `json:"sequence"`
	Membership string  `json:"membership"` // selected_live, selected_history, outside_selection
}

// The host records actual observed boundaries. Terminal commit and response
// finalization are separate; missing observations remain null, including crashes.
// Active owner review measurements are collected by the pilot, never inferred.
type CompilerForegroundMeasurement struct {
	RootID                    EventID `json:"root_id"`
	StartedAtUnixMS           int64   `json:"started_at_unix_ms"`
	TerminalCommittedAtUnixMS *int64  `json:"terminal_committed_at_unix_ms"`
	TerminalCommitNanos       *int64  `json:"terminal_commit_nanos"`
	ResponseFinalizedAtUnixMS *int64  `json:"response_finalized_at_unix_ms"`
	ResponseFinalizationNanos *int64  `json:"response_finalization_nanos"`
	Outcome                   string  `json:"outcome"`
}

// A selection unit's sequence range bounds its exact source members. It does
// not imply that interleaved events or interpretation context were evidence.
type CompilerDiagnosticUnit struct {
	SelectionID       string `json:"selection_id"`
	GenerationID      string `json:"generation_id"`
	JobID             string `json:"job_id,omitempty"`
	FirstSequence     int64  `json:"first_sequence"`
	LastSequence      int64  `json:"last_sequence"`
	State             string `json:"state"`
	Reason            string `json:"reason,omitempty"`
	SelectedNewEvents int64  `json:"selected_new_events"`
}
type CompilerDiagnosticRoot struct {
	ActivationID  string  `json:"activation_id"`
	RootID        EventID `json:"root_id"`
	FirstSequence int64   `json:"first_sequence"`
	LastSequence  int64   `json:"last_sequence"`
	State         string  `json:"state"`
	Reason        string  `json:"reason,omitempty"`
	SelectionID   string  `json:"selection_id,omitempty"`
}

File: .scratch/memory-stage-4/148-http-ui-frozen-files/internal/web/ui/src/api/compilerDiagnostics.ts
import { candidateReviewAPI, CandidateReviewError, type CandidateRef } from "./candidateReview";

export type DiagnosticView = "jobs" | "candidates" | "activations" | "history" | "selection" | "selections" | "live_roots" | "foreground";
export type DiagnosticQuery = { session_id: string; view: DiagnosticView; generation_id?: string; cursor?: string; limit: number };
export type DiagnosticSessions = { session_ids: string[]; next_cursor?: string };
export type AttemptMeasurement = { attempt: number; fence: number; claimed_at_unix_ms: number; queue_wait_nanos: number | null; inference_nanos: number | null; validation_nanos: number | null; database_completion_nanos: number | null; observed_outcome: string };
export type DiagnosticJob = { job_id: string; generation_id: string; session_id: string; first_sequence: number; last_sequence: number; state: string; reason?: string; attempts: number; retry_at?: number; pause_reason?: string; lane: string; recovery?: string; queued_at_unix_ms: number | null; published_at_unix_ms: number | null; selected_new_events: number; completed_new_events: number; candidate_freshness_nanos: number | null; publication_nanos: number | null; measurements: AttemptMeasurement[] };
export type DiagnosticCandidate = { ref: CandidateRef; job_id: string; generation_id: string; review_state: string; equivalent_to?: string; published_at_unix_ms: number | null; decided_at_unix_ms: number | null; edited: boolean };
export type DiagnosticActivation = { activation_id: string; selector: { source_scope: string; session_id?: string; destination: string }; generation_id: string; revision: number; after_position: number; through_position?: number; work_paused: boolean };
export type DiagnosticHistory = { request_id: string; range_index: number; generation_id: string; first_sequence: number; last_sequence: number; scanned_sequence: number; revision: number; cancelled: boolean };
export type DiagnosticUnit = { selection_id: string; generation_id: string; job_id?: string; first_sequence: number; last_sequence: number; state: string; reason?: string; selected_new_events: number };
export type DiagnosticRoot = { activation_id: string; root_id: string; first_sequence: number; last_sequence: number; state: string; reason?: string; selection_id?: string };
export type DiagnosticSelection = { event_id: string; sequence: number; membership: string };
export type ForegroundMeasurement = { root_id: string; started_at_unix_ms: number; terminal_committed_at_unix_ms: number | null; terminal_commit_nanos: number | null; response_finalized_at_unix_ms: number | null; response_finalization_nanos: number | null; outcome: string };
export type CompilerDiagnostics = { scope_key: string; session_id: string; view: DiagnosticView; as_of_unix_ms: number; revision: number; indexing: boolean; counts: Record<string, number>; capacity_state: string; jobs: DiagnosticJob[]; candidates: DiagnosticCandidate[]; activations: DiagnosticActivation[]; history: DiagnosticHistory[]; selection: DiagnosticSelection[]; foreground: ForegroundMeasurement[]; selections: DiagnosticUnit[]; live_roots: DiagnosticRoot[]; next_cursor?: string };

async function postDiagnostics<T>(path: string, scope_key: string, input: unknown): Promise<T> {
 const response = await fetch(`/api/memory/compiler/${path}`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ scope_key, input }), cache: "no-store" });
 const value = await response.json();
 if (!response.ok) throw new CandidateReviewError(value.code ?? "diagnostics_retryable", value.error ?? "Compiler diagnostics are unavailable. Refresh to try again.");
 return value as T;
}
export const compilerDiagnosticsAPI = {
 scopes: candidateReviewAPI.scopes,
 sessions: (scope: string, cursor = ""): Promise<DiagnosticSessions> => postDiagnostics("sessions", scope, { limit: 32, cursor }),
 inspect: (scope: string, query: DiagnosticQuery): Promise<CompilerDiagnostics> => postDiagnostics("diagnostics", scope, query),
};
export type CompilerDiagnosticsAPI = typeof compilerDiagnosticsAPI;
